File: web/dynamic_combo.js
import { app } from "../../scripts/app.js";

app.registerExtension({
  name: "MpiDynamicCombo",

  async beforeRegisterNodeDef(nodeType, nodeData) {
    if (nodeData.name !== "MpiDynamicCombo") return;

    const origOnNodeCreated = nodeType.prototype.onNodeCreated;

    nodeType.prototype.onNodeCreated = function () {
      if (origOnNodeCreated) origOnNodeCreated.call(this);

      const fields = [
        this.widgets.find(w => w.name === "option_1"),
        this.widgets.find(w => w.name === "option_2"),
        this.widgets.find(w => w.name === "option_3"),
        this.widgets.find(w => w.name === "option_4"),
        this.widgets.find(w => w.name === "option_5"),
        this.widgets.find(w => w.name === "option_6"),
        this.widgets.find(w => w.name === "option_7"),
        this.widgets.find(w => w.name === "option_8"),
      ];

      const combo = this.widgets.find(w => w.name === "selected");
      if (!combo) return;

      const updateComboChoices = () => {
        const choices = fields
          .map(w => w?.value?.trim())
          .filter(v => v); // remove empty

        const newChoices = choices.length ? choices : ["None"];
        const currentValue = combo.value;

        combo.options.values = newChoices;
        combo.value = newChoices.includes(currentValue) ? currentValue : newChoices[0];

        combo.inputEl?.dispatchEvent(new Event("change", { bubbles: true }));
        this.setDirtyCanvas(true);
      };

      // Bind callbacks early
      fields.forEach(w => {
        if (w) w.callback = updateComboChoices;
      });

      // Run once after full mount
      this.setup = async () => {
        updateComboChoices();
      };
    };
  }
});

// import { app } from "../../scripts/app.js";

// app.registerExtension({
//   name: "MpiDynamicCombo",

//   async beforeRegisterNodeDef(nodeType, nodeData) {
//     if (nodeData.name !== "MpiDynamicCombo") return;

//     const origOnNodeCreated = nodeType.prototype.onNodeCreated;

//     nodeType.prototype.onNodeCreated = function () {
//       if (origOnNodeCreated) origOnNodeCreated.call(this);

//       const fields = [
//         this.widgets.find(w => w.name === "option_1"),
//         this.widgets.find(w => w.name === "option_2"),
//         this.widgets.find(w => w.name === "option_3"),
//         this.widgets.find(w => w.name === "option_4"),
//         this.widgets.find(w => w.name === "option_5"),
//         this.widgets.find(w => w.name === "option_6"),
//         this.widgets.find(w => w.name === "option_7"),
//         this.widgets.find(w => w.name === "option_8"),
//       ];

//       const combo = this.widgets.find(w => w.name === "selected");
//       if (!combo) return;

//       const updateComboChoices = () => {
//         const choices = fields
//           .map(w => w?.value?.trim())
//           .filter(v => v); // remove empty

//         const newChoices = choices.length ? choices : ["None"];
//         const currentValue = combo.value;

//         combo.options.values = newChoices;
//         combo.value = newChoices.includes(currentValue) ? currentValue : newChoices[0];

//         combo.inputEl?.dispatchEvent(new Event("change", { bubbles: true }));
//         this.setDirtyCanvas(true);
//       };

//       fields.forEach(w => {
//         if (w) w.callback = updateComboChoices;
//       });

//       setTimeout(updateComboChoices, 50);
//     };
//   }
// });

File: dynamic_combo.py
class MpiDynamicCombo:
    @classmethod
    def INPUT_TYPES(cls):
        return {
            "required": {
                "option_1": ("STRING", {}),
                "option_2": ("STRING", {}),
                "option_3": ("STRING", {}),
                "option_4": ("STRING", {}),
                "option_5": ("STRING", {}),
                "option_6": ("STRING", {}),
                "option_7": ("STRING", {}),
                "option_8": ("STRING", {}),
                "selected": (
                    "COMBO",
                    {
                        "choices": ["None"],  # placeholder, will be overridden
                    },
                ),
            }
        }

    CATEGORY = "MpiNodes/Logic"
    RETURN_TYPES = ("STRING", "INT")
    FUNCTION = "use_selected"

    def use_selected(
        self,
        option_1,
        option_2,
        option_3,
        option_4,
        option_5,
        option_6,
        option_7,
        option_8,
        selected,
    ):
        options = [
            option_1,
            option_2,
            option_3,
            option_4,
            option_5,
            option_6,
            option_7,
            option_8,
        ]
        sel_int = next(
            (i + 1 for i, val in enumerate(options) if val == selected), -1
        )
        return (selected, sel_int)
